tests/unit/ModelType.cpp

```diff
@@ -674,3 +674,50 @@ BOOST_AUTO_TEST_CASE(Filters) {
   revng_check(NonScalarCount == 2);
   revng_check(ScalarCount + NonScalarCount == CFT.Arguments().size());
 }
+
+BOOST_AUTO_TEST_CASE(TypeComparison) {
+  // Basic sanity checks
+  using PT = model::PrimitiveType;
+  revng_check(PT::makeUnsigned(4) == PT::makeUnsigned(4));
+  revng_check(PT::makeUnsigned(4) != PT::makeSigned(4));
+  revng_check(PT::makeUnsigned(4) != PT::makeUnsigned(8));
+
+  model::Binary Model;
+
+  auto [FirstTD, FirstT] = Model.makeTypedefDefinition(PT::makeUnsigned(4));
+  revng_check(FirstTD == FirstT->asTypedef());
+
+  auto [SecondTD, SecondT] = Model.makeTypedefDefinition(PT::makeUnsigned(4));
+  revng_check(SecondTD == SecondT->asTypedef());
+
+  // Ensure different types are different, even if they only differ by the ID.
+  revng_check(FirstTD != SecondTD);
+  revng_check(FirstT != SecondT);
+  revng_check(FirstTD != SecondTD);
+
+  // Two distinct types are the same as long as they match completely.
+  revng_check(FirstT == Model.makeType(FirstTD.key()));
+  revng_check(SecondT != Model.makeType(FirstTD.key()));
+
+  // Even if they are wrapped in multiple layers.
+  auto FirstArray = model::ArrayType::make(Model.makeType(FirstTD.key()), 10);
+  auto SecondArray = model::ArrayType::make(Model.makeType(FirstTD.key()), 10);
+  revng_check(FirstArray == SecondArray);
+
+  // But not if those layers differ.
+  auto FirstPtr = model::PointerType::make(Model.makeType(FirstTD.key()), 8);
+  revng_check(FirstArray != FirstPtr);
+
+  // Consequently, maps and sets of types also work.
+  std::map<model::UpcastableType, int> Map;
+  FirstT->dump();
+  SecondT->dump();
```

Comment:
I'd avoid this

Reply:
Why? This makes the test failures really easy to debug.

If it's `dump` specifically you don't like, I don't mind doing `String += serializeToString(X)` dance. But I can't see why we'd want to.